### ISSUE TITLE
distrod doesn't run user.slice?

### ISSUE THREAD

**Issue by rockorequin** [nullpo-head/wsl-distrod#13]
After installing distrod, systemctl status only showed init.scope and system.slice running - there is no user.slice (and user-1000.slice, etc). Have I done something wrong? Or if not, there is a way perhaps to configure it to run the user systemd units? I was hoping to get gvfs-daemon running to allow nautilus to connect to sftp:// locations.

I installed distrod as per the instructions to add it to my existing distro, ie I downloaded and ran install.sh with the install argument and then ran "/opt/distrod/bin/distrod enable" (both as sudo, as the commands complain if not run as sudo). I then terminated wsl with "wsl --shutdown" (note that "wsl --terminate Distrod" as per the instructions didn't work, since I had only added distrod to my existing distro, so Distrod did not exist) and re-ran wsl.

**Comment by nullpo-head** [comment]
> there is no user.slice 

Yes, that's right. It is how Distrod runs today. I've been putting off implementing user slice because I didn't think there were that many use cases for it in WSL, where everyone has root. `gvfs` is something I missed and yes it looks reasonable use case. I add it to my backlog.

As a workaround, if you run `sudo login -f user_name` command once, a session under the user.slice will start. Or, you can do ssh into localhost.

**Comment by lungothrin** [comment]
+1

**Comment by rockorequin** [comment]
Thanks, looking forward to it.

Once distrod automatically starts user slice, does this mean that if I run wsl:vlc from the windows Start Menu that an sftp network link will work?

**Comment by nullpo-head** [comment]
I'm not exactly sure what you want to do because I'm not familiar with `wsl:vlc`, but apps launched from the start menus will also run under the systemd session and slices. So, if what you want depends on just systemd's user slices, it will successfully run. Of course I have to confirm by actually implementing it, though!

**Comment by rockorequin** [comment]
I was just using wsl:vlc as a short cut to saying "run vlc from wsl from the start menu". Sorry, it must have been confusing!

**Comment by nullpo-head** [comment]
@lopugit
> > Yes, this issue is a duplicate of #13. As a workaround, do sudo login -f username once. user slice will be implemented later.
>
> How often do I have to do this though? Just once ever or every session or?

To just trigger user systemd services, do it once per boot of WSL instance.
To use `systemctl --user`, do it when you want to use `systemctl --user`. In the session you started y `login`, you can use `systemctl --user`.

**Comment by lopugit** [comment]
> @lopugit
> 
> > > Yes, this issue is a duplicate of #13. As a workaround, do sudo login -f username once. user slice will be implemented later.
> > 
> > 
> > How often do I have to do this though? Just once ever or every session or?
> 
> To just trigger user systemd services, do it once per boot of WSL instance. To use `systemctl --user`, do it when you want to use `systemctl --user`. In the session you started y `login`, you can use `systemctl --user`.

Thanks a lot, really appreciate it

**Comment by sarim** [comment]
I'm not familiar with systemd internals, but is it possible that when distrod is creating the "bottle", it also runs `sudo login -f user_name` and then puts the shell sessions inside that?

**Comment by mgood7123** [comment]
has anyone gotten this to work?

i currently need to run `login -f archlinux` every time i log into wsl

with wsl conf
```
[user]
default=root
```
```

**Comment by joouha** [comment]
I found that enabling user-linger means that the user-session gets launched automatically when WSL boots:

`sudo loginctl enable-linger user_name`

**Comment by mgood7123** [comment]
> I found that enabling user-linger means that the user-session gets launched automatically when WSL boots:
> 
> `sudo loginctl enable-linger user_name`

Isn't root required o manipulate systemd

**Comment by joouha** [comment]
I don't think it will break systemd. Enabling linger means that the systemd user instance is started right after boot, and kept running after the last session closes.

This basically means that my user's `user-1000.slice` gets started automatically without the need to run `sudo login -f user_name` at every boot.

It's not how I'd do things on a desktop linux install, but for my use-case with WSL it suits me.

**Comment by mgood7123** [comment]
> @nullpo-head How can we add `PAMName` property when executing $SHELL? Adding that properly creates the shell with systemd user slice/session.
> 
> So long story short, I was doing a deep dive on how these systemd stuffs work and how to enable systemd user session. Geine uses this command `machinectl shell -q USERNAME@.host /bin/bash` to run shell. And it creates systemd user session. After looking through systemd source for how `machinectl shell` executes it, I found this:
> 
> https://github.com/systemd/systemd/blob/main/src/machine/machine-dbus.c#L732
> 
> The important property here is `PAMName`. I can replicate it with `systemd-run`.
> 
> So In a distrod configued wsl distro, I can execute the following commands to get a shell with systemd user session.
> 
> ```shell
> # from powershell start wsl to root user
> wsl -d UbuntuMain --user root
> # now in ubuntu, inside systemd thanks to distrod
> systemd-run --pty --service-type=simple --uid=gittu -p PAMName=login /bin/bash
> 
> # now in user session. Can run the followings succesfully
> env | grep XDG
> systemctl --user status
> ```
> 
> So can this be added to distrod? I don't know rust but looking at the code of distrod and geine, while geine executes commands like "machinectl daemonize unshare systemd" to do its job, distrod seems to be doing these programatically by api call? Is it a builtin or a rust library to interface with linux systemd / namespace etc... ? Maybe there is a option to add `PAMName=login` there?
> 
> Or you can maybe simply append `systemd-run --pty --service-type=simple --uid=USERNAME -p PAMName=login` to shell command?
> 
> Update:
> 
> I've created a small hack in go to run in user session: https://github.com/sarim/gbash

this works for ubuntu kenetic wsl2 windows 11

my original one fails, idk why

```
Welcome to Ubuntu Kinetic Kudu (development branch) (GNU/Linux 5.10.16.3-microsoft-standard-WSL2 x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

The programs included with the Ubuntu system are free software;

the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

-bash: cannot set terminal process group (28): Inappropriate ioctl for device
-bash: no job control in this shell

$ systemctl status --user
Failed to connect to bus: Permission denied
```

**Comment by mgood7123** [comment]
> @nullpo-head How can we add `PAMName` property when executing $SHELL? Adding that properly creates the shell with systemd user slice/session.
> 
> So long story short, I was doing a deep dive on how these systemd stuffs work and how to enable systemd user session. Geine uses this command `machinectl shell -q USERNAME@.host /bin/bash` to run shell. And it creates systemd user session. After looking through systemd source for how `machinectl shell` executes it, I found this:
> 
> https://github.com/systemd/systemd/blob/main/src/machine/machine-dbus.c#L732
> 
> The important property here is `PAMName`. I can replicate it with `systemd-run`.
> 
> So In a distrod configued wsl distro, I can execute the following commands to get a shell with systemd user session.
> 
> ```shell
> # from powershell start wsl to root user
> wsl -d UbuntuMain --user root
> # now in ubuntu, inside systemd thanks to distrod
> systemd-run --pty --service-type=simple --uid=gittu -p PAMName=login /bin/bash
> 
> # now in user session. Can run the followings succesfully
> env | grep XDG
> systemctl --user status
> ```
> 
> So can this be added to distrod? I don't know rust but looking at the code of distrod and geine, while geine executes commands like "machinectl daemonize unshare systemd" to do its job, distrod seems to be doing these programatically by api call? Is it a builtin or a rust library to interface with linux systemd / namespace etc... ? Maybe there is a option to add `PAMName=login` there?
> 
> Or you can maybe simply append `systemd-run --pty --service-type=simple --uid=USERNAME -p PAMName=login` to shell command?
> 
> Update:
> 
> I've created a small hack in go to run in user session: https://github.com/sarim/gbash

combined with this, it works for windows 11 ubuntu kinetic



`!!! WORKAROUND !!!`

https://github.com/nullpo-head/wsl-distrod/issues/64

create a `/etc/wsl.conf` with the contents
```
[user]
default=root
```

next install gbash to automatically start and wait for systemd at wsl boot
```
apt install git make golang
git clone https://github.com/sarim/gbash
cd gbash
make config
```

next add the following to `/etc/profile.d/distrod-user-wsl-envs.sh`
```sh
if [ "$(id -u)" == 0 ]; then
    systemd-run --pty --service-type=simple --uid=YOUR_USERNAME_HERE -p PAMName=login /bin/bash
fi
```

```
pspdev@LAPTOP-5PKRTSDE:/$ exit
exit
root@LAPTOP-5PKRTSDE:/mnt/c/Users/small# exit
logout
PS C:\Users\small> wsl --shutdown
PS C:\Users\small> wsl
Waiting for system socket
Waiting for system socket
Waiting for system socket
Waiting for system socket
Waiting for system socket
Waiting for system socket
Waiting for system socket
Waiting for system socket
Running as unit: run-u8.service
Press ^] three times within 1s to disconnect TTY.
pspdev@LAPTOP-5PKRTSDE:/$ systemctl status
Failed to dump process list for 'LAPTOP-5PKRTSDE', ignoring: Input/output error
● LAPTOP-5PKRTSDE
    State: running
    Units: 247 loaded (incl. loaded aliases)
     Jobs: 0 queued
   Failed: 0 units
    Since: Tue 2022-10-04 13:45:45 AEST; 16s ago
  systemd: 251.4-1ubuntu6
  Tainted: cgroupsv1
   CGroup: /
pspdev@LAPTOP-5PKRTSDE:/$ systemctl status --user
● LAPTOP-5PKRTSDE
    State: running
    Units: 104 loaded (incl. loaded aliases)
     Jobs: 0 queued
   Failed: 0 units
    Since: Tue 2022-10-04 13:45:46 AEST; 17s ago
  systemd: 251.4-1ubuntu6
  Tainted: cgroupsv1
   CGroup: /user.slice/user-1001.slice/user@1001.service
           └─init.scope
             ├─120 /lib/systemd/systemd --user
             └─122 "(sd-pam)"
pspdev@LAPTOP-5PKRTSDE:/$
```

**Comment by huangz986** [comment]
> > I found that enabling user-linger means that the user-session gets launched automatically when WSL boots:
> > `sudo loginctl enable-linger user_name`
> 
> That might break systemd, not sure

You are right, enable-linger for the user breaks the user's systemd, at least on my machine. You saved my time.
